packages/now-build-utils/src/detect-builders.ts

```diff
@@ -392,119 +553,412 @@ function validateFunctions(files: string[], { functions = {} }: Options) {
   return null;
 }
 
-// When zero config is used we can call this function
-// to determine what builders to use
-export async function detectBuilders(
-  files: string[],
-  pkg?: PackageJson | undefined | null,
-  options: Options = {}
-): Promise<{
-  builders: Builder[] | null;
-  errors: ErrorResponse[] | null;
-  warnings: ErrorResponse[];
-}> {
-  const errors: ErrorResponse[] = [];
-  const warnings: ErrorResponse[] = [];
+function checkUnusedFunctions(
+  frontendBuilder: Builder | null,
+  usedFunctions: Set<string>,
+  options: Options
+): ErrorResponse | null {
+  const unusedFunctions = new Set(
+    Object.keys(options.functions || {}).filter(key => !usedFunctions.has(key))
+  );
 
-  const functionError = validateFunctions(files, options);
+  if (!unusedFunctions.size) {
+    return null;
+  }
+
+  // Next.js can use functions only for `src/pages` or `pages`
+  if (frontendBuilder && frontendBuilder.use.startsWith('@now/next')) {
+    for (const fnKey of unusedFunctions.values()) {
+      if (fnKey.startsWith('pages/') || fnKey.startsWith('src/pages')) {
+        unusedFunctions.delete(fnKey);
+      } else {
+        return {
+          code: 'unused_function',
+          message: `The function for ${fnKey} can't be handled by any builder`,
+        };
+      }
+    }
+  }
+
+  if (unusedFunctions.size) {
+    const [unusedFunction] = Array.from(unusedFunctions);
 
-  if (functionError) {
     return {
-      builders: null,
-      errors: [functionError],
-      warnings,
+      code: 'unused_function',
+      message:
+        `The function for ${unusedFunction} can't be handled by any builder. ` +
+        `Make sure it is inside the api/ directory.`,
     };
   }
 
-  // Detect all builders for the `api` directory before anything else
-  const builders = detectApiBuilders(files, options);
-  const { projectSettings = {} } = options;
-  const { outputDirectory, buildCommand, framework } = projectSettings;
+  return null;
+}
 
-  let frontendBuilder: Builder | null = null;
+function getApiRoute(
+  fileName: string,
+  sortedFiles: string[],
+  options: Options
+): {
+  defaultRoute: Source | null;
+  dynamicRoute: Source | null;
+  routeError: ErrorResponse | null;
+} {
+  const conflictingSegment = getConflictingSegment(fileName);
+
+  if (conflictingSegment) {
+    return {
+      defaultRoute: null,
+      dynamicRoute: null,
+      routeError: {
+        code: 'conflicting_path_segment',
+        message:
+          `The segment "${conflictingSegment}" occurs more than ` +
+          `one time in your path "${fileName}". Please make sure that ` +
+          `every segment in a path is unique.`,
+      },
+    };
+  }
 
-  // Everything will be static if either is an empty string
-  const ignoreBuild = buildCommand === '' || outputDirectory === '';
+  const occurrences = pathOccurrences(fileName, sortedFiles).filter(
+    name => name !== fileName
```

Comment:
Let's add this filter into the `pathOccurrences` to avoid another loop and avoid testing files we don't need to.

It can be something like:

```js
if (file !== fileName) {
  if (absolutePath === currentAbsolutePath) {
      prev.push(file);
    } else if (partiallyMatches(filePath, file)) {
      prev.push(file);
    }
}
```

Reply:
Done 👍